File: LeetCode-Python/WordBreak.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
         # Kelime sözlüğünü hızlı aramalar için bir set'e dönüştür
        word_set = set(wordDict)
        
        n = len(s)
        # dp[i], s'nin ilk i karakterinin (s[0:i]) sözlükteki kelimelerle bölünüp bölünemeyeceğini gösterir.
        dp = [False] * (n + 1)
        
        # Boş string her zaman bölünebilir.
        dp[0] = True
        
        # Her bir uzunluk için kontrol et
        for i in range(1, n + 1):
            # String'in her olası bölüm noktasını kontrol et (j)
            for j in range(i):
                # Eğer dp[j] doğruysa (s[0:j] bölünebilir)
                # VE s[j:i] alt string'i sözlükte varsa, o zaman dp[i] doğrudur.
                if dp[j] and s[j:i] in word_set:
                    dp[i] = True
                    break # s[0:i] için bir segmentasyon bulduk, daha fazla kontrol etmeye gerek yok.
                    
        # s'nin tamamının (ilk n karakterinin) bölünüp bölünemediğini döndür.
        return dp[n]

File: LeetCode-Cpp/WordBreak.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
                // Sözlük kelimelerini hızlı arama için bir hash seti'ne kopyala
        std::unordered_set<std::string> dictSet(wordDict.begin(), wordDict.end());

        int n = s.length();
        // dp[i], s'nin ilk i karakterinin (s.substr(0, i)) sözlük kelimeleriyle segmentlere ayrılabildiğini gösterir.
        std::vector<bool> dp(n + 1, false);
        
        // Boş dizge her zaman segmentlere ayrılabilir.
        dp[0] = true;

        // Her bir uzunluk için kontrol et
        for (int i = 1; i <= n; ++i) {
            // Dizgenin her olası bölüm noktasını kontrol et
            for (int j = 0; j < i; ++j) {
                // Eğer dp[j] doğruysa (s'nin ilk j karakteri segmentlere ayrılabilir)
                // VE s[j...i-1] alt dizgesi sözlükte varsa, o zaman dp[i] doğrudur.
                if (dp[j] && dictSet.count(s.substr(j, i - j))) {
                    dp[i] = true;
                    break; // s[0...i-1] için bir segmentasyon bulduk, daha fazla kontrol etmeye gerek yok.
                }
            }
        }

        // s'nin tamamının (ilk n karakterinin) segmentlere ayrılıp ayrılamadığını döndür.
        return dp[n];
    }
};
